blog/bloom-and-cuckoo.md

```diff
@@ -0,0 +1,158 @@
+If a particular node does not store a requested piece of content, it can use the Bloom filters that it received from its peers to check if a peer stores the requested content. 
+We are using Counting Bloom filters rather than regular Bloom filters because we need to be able to remove elements from the filter to support cache eviction.
+
+## Bloom Filters
+A Bloom filter is a space-efficient probabilistic data structure that can be used to perform approximate set membership queries.
+The answer to an approximate set membership query is not `no` or `yes`, but rather `no` or `probably`. This `probably` is quantified with the false positive rate.  
+One of the convenient features of Bloom filters is that they can be configured to have a specific false positive rate.
+Of course there is a tradeoff here; the lower the false positive rate, the larger the memory footprint. Bloom filters support two operations: `insert` and `contains`.  
+A Bloom filter is represented by an array of $m$ bits together with $k$ independent hash functions. To insert an element into the filter, it is hashed with each of the $k$ hash functions.
+The resulting hashes are interpreted as integers (modulo $m$) to obtain $k$ array positions. The bits at these positions are then set to 1 (if there aren't already 1).  
+To check whether or not an element is contained in the filter, the element is hashed $k$ times with the different hash functions.
+If all bits at the resulting array positions are 1, the element is assumed to be present. If any of the $k$ bits are zero, we can be certain that the queried element is not present in the set.
+However, even if all bits are 1, it might still be the case that the bits were set by a combination of other elements. This is where the aforementioned false positive rate comes into play.  
+Since we also need a `remove` operation for our use case, we have been using Counting Bloom filters, a variant of Bloom filters.
+Counting Bloom filters retain most of the properties that regular Bloom filters have. The `remove` operation comes at the cost of an increased memory footprint.
+Each position in the array is no longer a single bit but rather a group of bits that represent a counter.
+Whenever an element is inserted into the filter, the counters for all $k$ positions are incremented by 1. To remove an element, we simply decrement the counters.
+
+## Cuckoo Filters
+Bloom filters are the most known members of a class of data structures called Approximate Membership Query Filters (AMQ Filters).
+A relatively recent addition to this class are Cuckoo filters [1]. Cuckoo filters share many similarities with Bloom filters, and especially Counting Bloom filters.
+They are space-efficient, they are used for approximate set membership queries, the false positive rate can be configured, and they support the operations `insert`, `contains`, and `remove`.  
```

Comment:
I'd break this sentence into two and structure like: 
"They are space-efficient and can be used for approximate set membership queries. Cuckoo filters also support the operations `insert`, `contains`, and `remove`, and have configurable false positive rates"